components/datetime/src/time_zone.rs

```diff
@@ -417,282 +740,457 @@ impl TimeZoneFormat {
+    Iso8601(IsoFormat, IsoMinutes, IsoSeconds), // -07:00
+}
+
+/// An enum for fallback formats.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum FallbackFormat {
+    Iso8601(IsoFormat, IsoMinutes, IsoSeconds),
+    LocalizedGmt,
+}
+
+/// A bag of options to define how time zone will be formatted.
+#[derive(Default, Debug, Clone, Copy, PartialEq)]
+pub struct TimeZoneFormatOptions {
+    pub fallback_format: Option<FallbackFormat>,
+}
+
+// Pacific Time
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub(super) struct GenericNonLocationLongFormat {}
+
+// PT
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub(super) struct GenericNonLocationShortFormat {}
+
+// Pacific Standard Time
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub(super) struct SpecificNonLocationLongFormat {}
+
+// PDT
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub(super) struct SpecificNonLocationShortFormat {}
+
+// Los Angeles Time
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub(super) struct GenericLocationFormat {}
+
+// GMT-07:00
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub(super) struct LocalizedGmtFormat {}
+
+// -07:00
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub(super) struct Iso8601Format {
+    format: IsoFormat,
+    minutes: IsoMinutes,
+    seconds: IsoSeconds,
+}
+
+// It is only used for pattern in special case and not public to users.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub(super) struct ExemplarCityFormat {}
+
+// An enum for time zone format unit.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub(super) enum TimeZoneFormatUnit {
+    GenericNonLocationLong(GenericNonLocationLongFormat),
+    GenericNonLocationShort(GenericNonLocationShortFormat),
+    SpecificNonLocationLong(SpecificNonLocationLongFormat),
+    SpecificNonLocationShort(SpecificNonLocationShortFormat),
+    GenericLocation(GenericLocationFormat),
+    LocalizedGmt(LocalizedGmtFormat),
+    Iso8601(Iso8601Format),
+    ExemplarCity(ExemplarCityFormat),
+}
+
+pub(super) trait FormatTimeZone {
+    fn format<W: fmt::Write + ?Sized>(
         &self,
         sink: &mut W,
         time_zone: &impl TimeZoneInput,
-    ) -> Result<(), DateTimeFormatError> {
-        // TODO(blocked on #277) Use formatter utility instead of replacing "{0}".
-        sink.write_str(
-            &self
-                .exemplar_cities
-                .as_ref()
-                .map(|p| p.get())
-                .and_then(|cities| time_zone.time_zone_id().and_then(|id| cities.0.get(id)))
-                .map(|location| {
-                    self.zone_formats
-                        .get()
-                        .region_format
-                        .replace("{0}", location)
-                })
-                .ok_or(fmt::Error)?,
+        data_payloads: &TimeZoneDataPayloads,
+    ) -> Result<fmt::Result, DateTimeFormatError>;
+
+    /// Formats a time segment with optional zero-padding.
+    fn format_time_segment(n: u8, padding: ZeroPadding) -> String {
```

Comment:
Nit: I think this is the wrong place to put this method. It should not be on the trait; the trait should have only the single `format` method. I think this method belongs as a private function on TimeZoneFormat?